components/tools/OmeroWeb/omeroweb/webgateway/views.py

```diff
@@ -2514,3 +2540,323 @@ def object_table_query(request, objtype, objid, conn=None, **kwargs):
+        """
+        Returns a response for failed login.
+        Reason for failure may be due to server 'error' or because
+        of form validation errors.
+
+        @param request:     http request
+        @param error:       Error message
+        @param form:        Instance of Login Form, populated with data
+        """
+        if error is None:
+            # If no error from server, maybe form wasn't valid
+            formErrors = []
+            for field in form:
+                for e in field.errors:
+                    formErrors.append("%s: %s" % (field.label, e))
+            error = " ".join(formErrors)
+        # Since @jsonp decorator can't return a 403,
+        # we do it manually. NB: this won't return jsonp 'callback()'
+        return JsonResponse({"message": error}, status=403)
+
+    def post(self, request, *args, **kwargs):
+        error = None
+        form = self.form_class(request.POST.copy())
+        if form.is_valid():
+            username = form.cleaned_data['username']
+            password = form.cleaned_data['password']
+            server_id = form.cleaned_data['server']
+            is_secure = form.cleaned_data['ssl']
+
+            connector = Connector(server_id, is_secure)
+
+            # TODO: version check should be done on the low level, see #5983
+            compatible = True
+            if settings.CHECK_VERSION:
+                compatible = connector.check_version(self.useragent)
+            if (server_id is not None and username is not None and
+                    password is not None and compatible):
+                conn = connector.create_connection(
+                    self.useragent, username, password,
+                    userip=get_client_ip(request))
+                # TODO: conn is None if user is INACTIVE (not in user group)...
+                if conn is not None:
+                    # Check if user is in "user" group
+                    roles = conn.getAdminService().getSecurityRoles()
+                    userGroupId = roles.userGroupId
+                    # ... so this will ALWAYS be True
+                    if userGroupId in conn.getEventContext().memberOfGroups:
+                        request.session['connector'] = connector
+                        # UpgradeCheck URL should be loaded from the server or
+                        # loaded omero.web.upgrades.url allows to customize web
+                        # only
+                        try:
+                            upgrades_url = settings.UPGRADES_URL
+                        except:
+                            upgrades_url = conn.getUpgradesUrl()
+                        upgradeCheck(url=upgrades_url)
+
+                        return self._handleLoggedIn(request, conn, connector)
+                    else:
+                        error = "This user is not active."
+                        return self._handleNotLoggedIn(self, request, error,
```

Comment:
I've removed that ```else``` statement since it's not possible to reach it now: It seems that ```conn``` is never returned for ```disabled``` users (not in user group). We'd need to catch inactive users deeper within the connection code. Added a TODO in PR description to look at this later. The additional login tests for webclient and api test various login failures & error messages.